text/0082-entity-tags.md

```diff
@@ -0,0 +1,251 @@
+};
+```
+
+This policy states that for a `User` to carry out the *writeDoc* action on a `Document`, the user must own the document, or else the user's job level must be at least 6 and the document and the user must each have a `write` tag, where at least one of the user's write-tag's values must be present in the document's write-tag's values.
+
+
+## Detailed design
+
+To support entity tags, we need to extend the JSON format for entities, the JSON and natural syntax for schemas, the policy syntax, as well as the evaluator, validator, and symbolic compiler.
+
+
+### Policy syntax and semantics
+
+Tags support two binary operators, where `e` and `k` are Cedar expressions:
+
+
+```
+e.hasTag(k)
+e.getTag(k)
+```
+
+
+The `hasTag` operator tests if the entity `e` has a tag with the key `k`.  This operator errors if `e` is not an entity, or `k` is not a string. It returns `false` if `e` doesn't have any tags at all to match the behavior of the `has` operator on entities and records. In other words, having no tags is equivalent to having an empty tags-map. The `getTag` operator retrieves the value for the key `k` from the tags for the entity `e`.  This operator errors under the same conditions as `hasTag` and additionally, if `e` has no tag with the key `k`.
+
+At the code level, we will extend the Cedar CST, AST, and EST to include the new operators (either as separate nodes in the relevant trees or as two extra binary operators), and we'll extend the evaluator to implement the above semantics.
+
+
+### JSON entity format
+
+The JSON entity format is extended to support optionally specifying tags separately from entity attributes:
+
+
+```
+[
+    {
+        "uid": {},
+        "parents": [],
+        "attrs": {},
+        "tags": {}
+    },
+    {
+        ...
+    }
+]
+```
+
+
+
+### Schema
+
+We extend schemas as follows to support the declaration of a tags. Here's the natural syntax:
+
+
+```
+Entity := 'entity' Idents ['in' EntOrTyps] [['='] EntityRecType] ['tags' Type] ';'
+```
+
+
+The new syntax simply extends the `Entity` production to enable optionally specifying the type of attached tags.
+
+The JSON syntax for schemas specifies tags as a separate `tags` field that specifies the type of the tag values. Here's our introductory example schema in this format:
+
+```
+"entityTypes": {
+    "User" : {
+        "shape" : {
+            "type" : "Record",
+            "attributes" : {
+                "jobLevel" : {
+                    "type" : "Long"
+                },
+            }
+        },
+        "tags" : {
+            "type" : "Set",
+            "element": { "type": "String" }
+        }
+    },
+    "Document" : {
+        "shape" : {
+            "type" : "Record",
+            "attributes" : {
+                "owner" : {
+                    "type" : "Entity",
+                    "name" : "User"
+                },
+            }
+        },
+        "tags" : {
```

Comment:
Can `tags` be on a `commonType`? If so, how do nested tags work? (I'm probably getting the syntax wrong) Ex:
```json
{
  "entityTypes": {
    "Document": {
      "shape": {
        "type": "Record",
        "attributes": {
          "name": {"type": "String"},
          "signature": {"type": "Signature"}
        },
      },
      "tags": {
        //...
      }
    }
  },
  "commonTypes": {
    "Signature": {
      "shape": {
        //...
      },
      "tags": {
        //..
      }
    }
  }
}
```

Reply:
Ok, after reading https://github.com/cedar-policy/rfcs/pull/82/files#r1759501086, this looks to be possible.

```
permit(principal is User, action, resource is apps::Deployment)
when {
  principal.hasTag("app") &&
  resource.hasTag("app") && 
  resource.spec.template.hasTag("app") &&
  principal.getTag("app").contains(resource.getTag("app")) &&
  principal.getTag("app").contains(resource.spec.template.getTag("app"))
};
```

Reply:
I'm not quite sure I understand the question, but it's possible to nest tags arbitrarily using the indirection approach described in the RFC (see `IntTags` and `StringTags` entities).